include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+  DALI_HOST_DEV
+  constexpr int compare(const TensorLayout &tl) const noexcept {
+    int n = ndim() < tl.ndim() ? ndim() : tl.ndim();
+    // <= to include the null terminator
+    for (int i = 0; i <= n; i++) {
+      int d = data_[i] - tl.data_[i];
+      if (d)
+        return d;
+    }
+    return 0;
+  }
+
+  DALI_HOST_DEV
+  bool operator<(const TensorLayout &tl) const noexcept {
+    return compare(tl) < 0;
+  }
+  DALI_HOST_DEV
+  bool operator>(const TensorLayout &tl) const noexcept {
+    return compare(tl) > 0;
+  }
+  DALI_HOST_DEV
+  bool operator<=(const TensorLayout &tl) const noexcept {
+    return compare(tl) <= 0;
+  }
+  DALI_HOST_DEV
+  bool operator>=(const TensorLayout &tl) const noexcept {
+    return compare(tl) >= 0;
+  }
+  DALI_HOST_DEV
+  bool operator==(const TensorLayout &tl) const noexcept {
+    return data_[max_ndim] == tl.data_[max_ndim] && compare(tl) == 0;
+  }
+  DALI_HOST_DEV
+  bool operator!=(const TensorLayout &tl) const noexcept {
+    return !(*this == tl);
+  }
+
+  /** @brief Number of characters, excluding the (always present) null terminator */
+  DALI_HOST_DEV
+  constexpr uint8_t size() const noexcept { return max_ndim - data_[max_ndim]; }
+  /** @brief Number of dimensions described by this object; same value as size() */
+  DALI_HOST_DEV
+  constexpr int ndim() const noexcept { return size(); }
+  /** @brief Returns true if size() == 0 */
+  DALI_HOST_DEV
+  constexpr bool empty() const noexcept { return size() == 0; }
+
+  DALI_HOST_DEV
+  bool is_permutation_of(TensorLayout b) const {
+    // argument passed by value, because we'll reorder it to match *this
+    if (ndim() != b.ndim())
+      return false;
+
+    int n = ndim();
+    // not the nicest O(n^2) algorithm, but who cares with ndim() <= 15
+    for (int i = 0; i < n; i++) {
+      char c = data_[i];
+      if (c == b[i])
+          continue;
+      int j;
+      for (j = i + 1; j < n; j++) {
+        if (b[j] == c)
+          break;
+      }
+      if (j == n)
+        return false;
+      char tmp = b[i];
+      b[i] = b[j];
+      b[j] = tmp;
+    }
+    return true;
+  }
+
+  using iterator = char*;
+  using const_iterator = const char*;
+
+  DALI_HOST_DEV
+  constexpr iterator begin() noexcept               { return data_; }
+  DALI_HOST_DEV
+  constexpr iterator end() noexcept                 { return data_ + ndim(); }
+  DALI_HOST_DEV
+  constexpr auto begin() const noexcept             { return cbegin(); }
+  DALI_HOST_DEV
+  constexpr auto end() const noexcept               { return cend(); }
+  DALI_HOST_DEV
+  constexpr const_iterator cbegin() const noexcept  { return data_; }
+  DALI_HOST_DEV
+  constexpr const_iterator cend() const noexcept    { return data_ + ndim(); }
+
+  DALI_HOST_DEV
+  TensorLayout sample_layout() const {
+    if (empty())
+      return {};
+#ifdef __CUDA_ARCH__
+    assert(data_[0] == 'N');
+#else
+    if (find('N') != 0)
```

Comment:
Lets enforce the same thing here: `DALI_ENFORCE(data_[0] == 'N', ...)`, otherwise `sub(1)` doesn't make any sense